File: frontend/src/types/backtest.ts
// 백테스트 관련 TypeScript 타입 정의

export type BacktestStatus =
    | 'QUEUED'
    | 'INITIALIZING'
    | 'DATA_COLLECTION'
    | 'SIGNAL_GENERATION'
    | 'SIMULATION'
    | 'ANALYSIS'
    | 'COMPLETED'
    | 'FAILED'
    | 'CANCELLED';

export interface BacktestConfig {
    name: string;
    description?: string;
    watchlist_id: string;
    strategy_id: string;
    start_date: string;
    end_date: string;
    initial_capital: number;
    commission: number;
    slippage: number;
    position_sizing: 'equal_weight' | 'market_cap' | 'volatility_adjusted';
    rebalancing_frequency: 'daily' | 'weekly' | 'monthly' | 'quarterly';
    risk_management?: RiskManagementConfig;
}

export interface RiskManagementConfig {
    max_position_size: number;
    stop_loss?: number;
    take_profit?: number;
    max_drawdown_limit?: number;
}

export interface BacktestSummary {
    id: string;
    name: string;
    status: BacktestStatus;
    strategy_name: string;
    strategy_type: string;
    total_return?: number;
    sharpe_ratio?: number;
    max_drawdown?: number;
    created_at: string;
    updated_at: string;
    completed_at?: string;
    duration?: number;
    is_favorite: boolean;
    tags: string[];
}

export interface BacktestDetail extends BacktestSummary {
    description?: string;
    watchlist_name: string;
    watchlist_symbols: string[];
    config: BacktestConfig;
    results?: BacktestResults;
    executions: BacktestExecution[];
}

export interface BacktestResults {
    total_return: number;
    annual_return: number;
    sharpe_ratio: number;
    max_drawdown: number;
    volatility: number;
    win_rate: number;
    total_trades: number;
    profit_factor: number;
    sortino_ratio?: number;
    calmar_ratio?: number;
    var_95?: number;
    cvar_95?: number;
    performance_data: PerformanceDataPoint[];
    trade_history: Trade[];
    monthly_returns: MonthlyReturn[];
}

export interface PerformanceDataPoint {
    date: string;
    portfolio_value: number;
    benchmark_value?: number;
    drawdown: number;
    returns: number;
}

export interface Trade {
    id: string;
    date: string;
    symbol: string;
    side: 'BUY' | 'SELL';
    quantity: number;
    price: number;
    commission: number;
    pnl: number;
    cumulative_pnl: number;
}

export interface MonthlyReturn {
    month: string;
    return: number;
    benchmark_return?: number;
}

export interface BacktestExecution {
    id: string;
    backtest_id: string;
    status: BacktestStatus;
    started_at: string;
    completed_at?: string;
    error_message?: string;
    progress: ExecutionProgress;
    logs: ExecutionLog[];
    intermediate_results?: IntermediateResults;
}

export interface ExecutionProgress {
    current_step: string;
    total_steps: number;
    completed_steps: number;
    percentage: number;
    estimated_completion?: string;
}

export interface ExecutionLog {
    timestamp: string;
    level: 'INFO' | 'WARNING' | 'ERROR' | 'DEBUG';
    message: string;
    details?: Record<string, any>;
}

export interface IntermediateResults {
    data_points_processed: number;
    signals_generated: number;
    trades_executed: number;
    current_portfolio_value?: number;
    current_return?: number;
}

export interface BacktestFilters {
    status?: BacktestStatus[];
    strategy_types?: string[];
    date_range?: {
        start: string;
        end: string;
    };
    performance_range?: {
        min_return?: number;
        max_return?: number;
    };
    tags?: string[];
    favorites_only?: boolean;
    search?: string;
}

export interface BacktestListParams extends BacktestFilters {
    page?: number;
    limit?: number;
    sort_by?: 'created_at' | 'updated_at' | 'total_return' | 'name';
    sort_order?: 'asc' | 'desc';
}

export interface BacktestListResponse {
    backtests: BacktestSummary[];
    total: number;
    page: number;
    pages: number;
    has_next: boolean;
    has_prev: boolean;
}

export interface CreateBacktestResponse {
    backtest_id: string;
    status: BacktestStatus;
    estimated_duration?: number;
    queue_position?: number;
}

export interface BacktestUpdate {
    name?: string;
    description?: string;
    is_favorite?: boolean;
    tags?: string[];
}

// API 응답 타입
export interface BacktestEstimateResponse {
    estimated_duration: number; // seconds
    estimated_cost?: number;
    resource_requirements: {
        cpu: number;
        memory: number;
        storage: number;
    };
}

// 프리셋 설정 타입
export interface BacktestPreset {
    id: string;
    name: string;
    description: string;
    config: Partial<BacktestConfig>;
    is_default: boolean;
}

// 비교 분석 타입
export interface BacktestComparison {
    backtests: BacktestSummary[];
    correlation_matrix: number[][];
    performance_comparison: PerformanceComparison[];
}

export interface PerformanceComparison {
    metric: string;
    values: Record<string, number>; // backtest_id -> value
    best_performer: string;
    worst_performer: string;
}

// 상태 관리 타입
export interface BacktestState {
    // 목록 관리
    backtests: BacktestSummary[];
    filters: BacktestFilters;
    loading: boolean;
    error?: string;

    // 실행 상태
    activeExecutions: Record<string, BacktestExecution>;

    // 선택된 백테스트
    selectedBacktest?: BacktestDetail;

    // 비교 모드
    comparisonMode: boolean;
    selectedForComparison: string[];

    // UI 상태
    viewMode: 'card' | 'table';
    showFilters: boolean;
}

export interface BacktestActions {
    // 목록 관리
    fetchBacktests: (params?: BacktestListParams) => Promise<void>;
    updateFilters: (filters: Partial<BacktestFilters>) => void;
    clearFilters: () => void;

    // 백테스트 실행
    createBacktest: (config: BacktestConfig) => Promise<string>;
    estimateBacktest: (config: Partial<BacktestConfig>) => Promise<BacktestEstimateResponse>;
    cancelBacktest: (id: string) => Promise<void>;

    // 상세 조회
    fetchBacktestDetail: (id: string) => Promise<void>;
    subscribeToExecution: (id: string) => void;
    unsubscribeFromExecution: (id: string) => void;

    // 관리 기능
    toggleFavorite: (id: string) => Promise<void>;
    updateTags: (id: string, tags: string[]) => Promise<void>;
    deleteBacktest: (id: string) => Promise<void>;
    cloneBacktest: (id: string) => Promise<string>;

    // 비교 기능
    toggleComparison: (id: string) => void;
    clearComparison: () => void;
    compareBacktests: (ids: string[]) => Promise<BacktestComparison>;

    // UI 액션
    setViewMode: (mode: 'card' | 'table') => void;
    toggleFilters: () => void;
}

// 폼 관련 타입
export interface BacktestFormData extends Omit<BacktestConfig, 'watchlist_id' | 'strategy_id'> {
    watchlist: { id: string; name: string } | null;
    strategy: { id: string; name: string; type: string } | null;
}

export interface BacktestFormErrors {
    name?: string;
    watchlist?: string;
    strategy?: string;
    start_date?: string;
    end_date?: string;
    initial_capital?: string;
    commission?: string;
    slippage?: string;
}

File: backend/app/schemas/backtest.py
"""
Backtest API Schemas
"""

from datetime import datetime
from typing import Any

from pydantic import Field
from .base_schema import BaseSchema
from app.models.backtest import (
    BacktestConfig,
    BacktestStatus,
    PerformanceMetrics,
    Position,
    Trade,
)


# Request Schemas
class BacktestCreate(BaseSchema):
    """백테스트 생성 요청"""

    name: str = Field(..., description="백테스트 이름")
    description: str = Field(default="", description="백테스트 설명")
    config: BacktestConfig = Field(..., description="백테스트 설정")


class BacktestUpdate(BaseSchema):
    """백테스트 수정 요청"""

    name: str | None = Field(None, description="백테스트 이름")
    description: str | None = Field(None, description="백테스트 설명")
    config: BacktestConfig | None = Field(None, description="백테스트 설정")


class BacktestExecutionRequest(BaseSchema):
    """백테스트 실행 요청"""

    signals: list[dict[str, Any]] = Field(..., description="트레이딩 시그널 목록")

    class Config:
        json_schema_extra = {
            "example": {
                "signals": [
                    {
                        "symbol": "AAPL",
                        "action": "BUY",
                        "quantity": 10,
                        "timestamp": "2023-01-01T00:00:00",
                    },
                    {
                        "symbol": "AAPL",
                        "action": "SELL",
                        "quantity": 5,
                        "timestamp": "2023-01-02T00:00:00",
                    },
                ]
            }
        }


# Response Schemas
class BacktestResponse(BaseSchema):
    """백테스트 응답"""

    id: str = Field(..., description="백테스트 ID")
    name: str = Field(..., description="백테스트 이름")
    description: str = Field(..., description="백테스트 설명")
    config: BacktestConfig = Field(..., description="백테스트 설정")
    status: BacktestStatus = Field(..., description="실행 상태")
    start_time: datetime | None = Field(None, description="실행 시작 시간")
    end_time: datetime | None = Field(None, description="실행 종료 시간")
    duration_seconds: float | None = Field(None, description="실행 시간(초)")
    performance: PerformanceMetrics | None = Field(None, description="성과 지표")
    created_at: datetime = Field(..., description="생성 시간")
    updated_at: datetime | None = Field(None, description="수정 시간")


class BacktestListResponse(BaseSchema):
    """백테스트 목록 응답"""

    backtests: list[BacktestResponse] = Field(..., description="백테스트 목록")
    total: int = Field(..., description="총 개수")


class BacktestExecutionResponse(BaseSchema):
    """백테스트 실행 응답"""

    id: str = Field(..., description="실행 ID")
    backtest_id: str = Field(..., description="백테스트 ID")
    execution_id: str = Field(..., description="실행 ID")
    start_time: datetime = Field(..., description="실행 시작 시간")
    end_time: datetime | None = Field(None, description="실행 종료 시간")
    status: BacktestStatus = Field(..., description="실행 상태")
    portfolio_values: list[float] = Field(..., description="포트폴리오 가치 히스토리")
    trades: list[Trade] = Field(..., description="거래 내역")
    positions: dict[str, Position] = Field(..., description="최종 포지션")
    error_message: str | None = Field(None, description="오류 메시지")
    created_at: datetime = Field(..., description="생성 시간")


class BacktestExecutionListResponse(BaseSchema):
    """백테스트 실행 목록 응답"""

    executions: list[BacktestExecutionResponse] = Field(..., description="실행 목록")
    total: int = Field(..., description="총 개수")


class BacktestResultResponse(BaseSchema):
    """백테스트 결과 응답"""

    id: str = Field(..., description="결과 ID")
    backtest_id: str = Field(..., description="백테스트 ID")
    execution_id: str = Field(..., description="실행 ID")

    # 성과 지표
    total_return: float = Field(..., description="총 수익률")
    annualized_return: float = Field(..., description="연환산 수익률")
    volatility: float = Field(..., description="변동성")
    sharpe_ratio: float = Field(..., description="샤프 비율")
    max_drawdown: float = Field(..., description="최대 낙폭")

    # 추가 지표
    calmar_ratio: float | None = Field(None, description="칼마 비율")
    sortino_ratio: float | None = Field(None, description="소르티노 비율")

    # 벤치마크 비교
    benchmark_return: float | None = Field(None, description="벤치마크 수익률")
    alpha: float | None = Field(None, description="알파")
    beta: float | None = Field(None, description="베타")

    # 거래 통계
    total_trades: int = Field(..., description="총 거래 수")
    winning_trades: int = Field(..., description="승리 거래 수")
    losing_trades: int = Field(..., description="패배 거래 수")
    win_rate: float = Field(..., description="승률")

    # 데이터 경로
    portfolio_history_path: str | None = Field(None, description="포트폴리오 히스토리 파일 경로")
    trades_history_path: str | None = Field(None, description="거래 히스토리 파일 경로")

    created_at: datetime = Field(..., description="생성 시간")


class BacktestResultListResponse(BaseSchema):
    """백테스트 결과 목록 응답"""

    results: list[BacktestResultResponse] = Field(..., description="결과 목록")
    total: int = Field(..., description="총 개수")


# Integrated Backtest Schemas
class IntegratedBacktestRequest(BaseSchema):
    """통합 백테스트 요청"""

    name: str = Field(..., description="백테스트 이름")
    description: str = Field(default="", description="백테스트 설명")
    symbols: list[str] = Field(..., description="심볼 목록")
    start_date: datetime = Field(..., description="시작일")
    end_date: datetime = Field(..., description="종료일")
    strategy_type: str = Field(..., description="전략 타입")
    strategy_params: dict[str, Any] = Field(default_factory=dict, description="전략 매개변수")
    initial_capital: float = Field(default=100000.0, description="초기 자본")


class IntegratedBacktestResponse(BaseSchema):
    """통합 백테스트 응답"""

    backtest_id: str = Field(..., description="백테스트 ID")
    execution_id: str | None = Field(None, description="실행 ID")
    result_id: str | None = Field(None, description="결과 ID")
    status: BacktestStatus = Field(..., description="상태")
    message: str = Field(..., description="메시지")
    performance: PerformanceMetrics | None = Field(None, description="성과 지표")
    start_time: datetime | None = Field(None, description="시작 시간")
    end_time: datetime | None = Field(None, description="종료 시간")
